include/swift/SIL/Notifications.h

```diff
@@ -13,9 +13,210 @@
+
+  /// Observe that we deserialized a witness-table declaration.
+  virtual void didDeserialize(ModuleDecl *mod,
+                              SILWitnessTable *wtable) override {}
+
+  /// Observe that we deserialized a default witness-table declaration.
+  virtual void didDeserialize(ModuleDecl *mod,
+                              SILDefaultWitnessTable *wtable) override {}
+
+  virtual StringRef getName() const = 0;
+
+  virtual ~DeserializationNotificationHandler() = default;
+};
+
+/// A notification handler that only overrides didDeserializeFunctionBody and
+/// calls the passed in function pointer.
+class FunctionBodyDeserializationNotificationHandler final
+    : public DeserializationNotificationHandler {
+public:
+  using Handler = void (*)(ModuleDecl *, SILFunction *);
+
+private:
+  Handler handler;
+
+public:
+  FunctionBodyDeserializationNotificationHandler(Handler handler)
+      : handler(handler) {}
+  virtual ~FunctionBodyDeserializationNotificationHandler() {}
+
+  void didDeserializeFunctionBody(ModuleDecl *mod, SILFunction *fn) override {
+    (*handler)(mod, fn);
+  }
+
+  StringRef getName() const override {
+    return "FunctionBodyDeserializationNotificationHandler";
+  }
+};
+
+/// A type that enables us to both pass in custom notification handler types
+/// that use a unique_ptr as well as pass managers and analyses that we want
+/// to pass in a bare pointer for. We use the API
+class DeserializationNotificationHandlerSet
+    : public DeserializationNotificationHandlerBase {
+public:
+  using NotificationUniquePtr =
+    std::unique_ptr<DeserializationNotificationHandler>;
+
+private:
+  /// A list of deserialization callbacks that update the SILModule and other
+  /// parts of SIL as deserialization occurs.
+  ///
+  /// We use 3 here since that is the most that will ever be used today in the
+  /// compiler. If that changed, that number should be changed as well. The
+  /// specific users are:
+  ///
+  /// 1. SILModule's serialization callback.
+  /// 2. SILPassManager notifications.
+  /// 3. Access Enforcement Stripping notification.
+  SmallVector<NotificationUniquePtr, 3> handlerSet;
+
+public:
+  DeserializationNotificationHandlerSet() = default;
+  ~DeserializationNotificationHandlerSet() = default;
+
+  void erase(DeserializationNotificationHandler *handler) {
+    auto iter = find_if(handlerSet, [&](const NotificationUniquePtr &h) {
+      return handler == h.get();
+    });
+    if (iter == handlerSet.end())
+      return;
+    handlerSet.erase(iter);
+  }
+
+  void add(NotificationUniquePtr &&handler) {
+    handlerSet.emplace_back(std::move(handler));
+    sortUnique(handlerSet);
+  }
+
+  static DeserializationNotificationHandler *
+  getUnderlyingHandler(const NotificationUniquePtr &h) {
+    return h.get();
+  }
+
+  using iterator = llvm::mapped_iterator<
+      decltype(handlerSet)::const_iterator,
+      decltype(&DeserializationNotificationHandlerSet::getUnderlyingHandler)>;
```

Comment:
Nice. Although as a user I'm always going to dig around to be sure of the type returned by dereferencing the iterator. On methods that return an iterator, it might be helpful to comment the iterator's value type.